### PR TITLE
Include more class in documentation

### LINKED ISSUE #17770: Some build items are missing in documentation
## Describe the bug
The Jbang script generating the https://quarkus.io/guides/all-builditems documentation overview page currently only includes build items which are defined in source files ending in `BuildItem.java` (see https://github.com/quarkusio/quarkus/blob/adf1beb64b0196039fd0216313e43e68ce3f3ad2/.github/quarkusbuilditemdoc.java#L96). There are however also build items which either don't adhere to this convention (e.g. `Capabilities`) or are defined as nested classes (e.g. in the processor class). As a result some build items are missing in the generated overview page.

### Expected behavior
All (public?) build items should be included in the generated overview page.

### Actual behavior
See description.

### PR DESCRIPTION
**Pull request by delongmeng**
Simply checking if the java file contains strings like "extends BuildItem {", "extends SimpleBuildItem {", etc.
This extends the build items to include build items not ending with "BuildItem", such as "Capabilities.java", "CapabilityAggregationStep.java", and "ProcessInheritIODisabled.java".

Fixes: #17770

### CODE REVIEW

**Review by gastaldi — COMMENTED** [review]
That makes sense. I added some comments.

Another solution is to check if `source.getSuperType()` is one of those types from https://github.com/quarkusio/quarkus/blob/main/.github/quarkusbuilditemdoc.java#L107

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [inline .github/quarkusbuilditemdoc.java:104]: Since the file is already read in memory, you can pass those contents to the `Roaster.parse` method

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:103]: yes I can change it back. I commented out to compare the output between the original and modified code :)

Comment (gastaldi) [reply .github/quarkusbuilditemdoc.java:103]: Don' t worry about that, the GitHub UI already does that for you 😉

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:99]: Do you mean line 99? I can try to change it.

Comment (gastaldi) [reply .github/quarkusbuilditemdoc.java:99]: > Do you mean line 99? I can try to change it.

Yes, you can create that only once, no need to create that list everytime a file is found

Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:99]: right! that makes sense!

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [inline .github/quarkusbuilditemdoc.java:102]: Remove this line

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [inline .github/quarkusbuilditemdoc.java:95]: Actually I meant to move it to a private static final constant. I think you can also remove the `{` from each item too. 
```suggestion
        private static final List<String> KEYWORDS = List.of("extends BuildItem", "extends SimpleBuildItem",
                "extends MultiBuildItem", "extends EmptyBuildItem");
```

**Review by gastaldi — CHANGES_REQUESTED** [review]

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [inline .github/quarkusbuilditemdoc.java:103]: No need to comment that, it' s in the Git history already 😉 
```suggestion
                    if (path.toString().endsWith("BuildItem.java") || match) {
```

Comment (gastaldi) [reply .github/quarkusbuilditemdoc.java:103]: Also I think we don't care anymore if the path ends with `BuildItem.java`. Knowing that it extends from those types may probably be enough

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:102]: sure

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:95]: Actually I included " {" here in purpose. I found that `quarkus/deployment/ExtensionLoader.java` will be picked up as well because of the code on line 273. It includes "extends SimpleBuildItem" but I don't think ExtensionLoader.java should be recognized as a build item? (correct me if I'm wrong)

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java]: @gastaldi I've made the change. Thank you so much for detailed suggestions!

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [inline .github/quarkusbuilditemdoc.java:94]: Move this to a class attribute

**Comment by gastaldi** [comment]
Can you squash all your commits into a single one? That would be necessary before merging.

Thanks!

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [reply .github/quarkusbuilditemdoc.java]: Oh, I didn't know about that, in this case perhaps it's better to keep the `{`, or https://github.com/quarkusio/quarkus/pull/22011#pullrequestreview-826449086, WDYT?

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:94]: Do you mean put it after line 49?

Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:94]: Now I have an error : "org.jboss.forge.roaster.ParserException:  Source does not represent a [JavaClassSource], instead was [JavaAnnotationImpl] - Cannot convert."

Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:94]: seem to be from this line: 
JavaClassSource source = Roaster.parse(JavaClassSource.class, fileContent);
I'm trying to resolve it

Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:94]: It still works after I added back " {" to the keywords

**Comment by delongmeng** [comment]
Yes let me see how I squash commits

**Comment by delongmeng** [comment]
> Can you squash all your commits into a single one? That would be necessary before merging.
> 
> Thanks!

Yes let me see how I squash commits

**Comment by delongmeng** [comment]
> > Can you squash all your commits into a single one? That would be necessary before merging.
> > Thanks!
> 
> Yes let me see how I squash commits

I think it worked @gastaldi

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [reply .github/quarkusbuilditemdoc.java]: Ah shoot, yeah, I forgot that not every .java file in this code is a class, my bad

File: .github/quarkusbuilditemdoc.java
Comment (gastaldi) [inline .github/quarkusbuilditemdoc.java:22]: This import can be removed

Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:22]: OK sure!

**Review by gastaldi — DISMISSED** [review]

File: .github/quarkusbuilditemdoc.java
Comment (delongmeng) [reply .github/quarkusbuilditemdoc.java:22]: But then do I need to squash again?

**Comment by geoand** [comment]
Can you please squash the commits?

Thanks

**Review by gsmet — CHANGES_REQUESTED** [review]
Do we know if it will slow down the build significantly? Because AFAICS, we are now reading all the files and that could be very time consuming?

Until we have a better knowledge of this, I would like us to not merge it.

**Comment by gastaldi** [comment]
> Do we know if it will slow down the build significantly? Because AFAICS, we are now reading all the files and that could be very time consuming?
> 
> Until we have a better knowledge of this, I would like us to not merge it.

That's a good point, perhaps it's easier to hardcode the missing items here instead (the ones that do not end with `BuildItem.java`)

**Comment by gastaldi** [comment]
In parallel, I think we should also consider renaming classes that do not follow this pattern (eg. `ProcessInheritIODisabled`to `ProcessInheritIODisabledBuildItem`)

**Comment by gsmet** [comment]
@gastaldi yeah I think you're on the right track. I would special case the few that make sense and we want to keep and fix the others. I don't have time to drive this so I'll let you see with the OP :)

**Comment by gastaldi** [comment]
@delongmeng do you have a list of the extra build items covered by this change?

**Review by gastaldi — CHANGES_REQUESTED** [review]
Just realized that this is not going to work because inner classes also contain `extends XXXBuildItem {` and they are not used (or supposed to be used) in the process method

**Comment by delongmeng** [comment]
> @delongmeng do you have a list of the extra build items covered by this change?

I saw three additional files: Capabilities, CapabilityAggregationStep, and ProcessInheritIODisabled, when I limited the directories to /core/deployment  and /core/test-extension.

Saw the conversations above. Yes if this doesn't affect a large scope of files it might be easier to just follow the naming convention.

**Comment by gastaldi** [comment]
I don't think `CapabilityAggregationStep` is intended to be used in other implementations. Given that [`ProcessInheritIODisabled`is now deprecated in favor of `ProcessInheritIODisabledBuildItem`,](https://github.com/quarkusio/quarkus/pull/22081)  `Capabilities` seems to be the only missing item. Can you hardcode that in the class instead?

**Comment by delongmeng** [comment]
> I don't think `CapabilityAggregationStep` is intended to be used in other implementations. Given that [`ProcessInheritIODisabled`is now deprecated in favor of `ProcessInheritIODisabledBuildItem`,](https://github.com/quarkusio/quarkus/pull/22081) `Capabilities` seems to be the only missing item. Can you hardcode that in the class instead?

Got it. I can definitely hardcode that :)

**Comment by gsmet** [comment]
@delongmeng can you have another look at that? Thanks!

**Comment by gsmet** [comment]
Closing this one as we will need a simpler patch. If you still want to work on this, let's do it in a new PR.

**Pull request closed without merging**
